quinn-h3/src/streams.rs

```diff
@@ -83,56 +82,74 @@ impl Future for RecvUni {
 
 pub struct PushStream(FrameStream);
 
-pub struct SendControlStream {
-    conn: ConnectionRef,
-    state: SendControlStreamState,
-    send: Option<SendStream>,
+pub struct SendUni {
+    ty: StreamType,
+    state: SendUniState,
+    data: VecDeque<Bytes>,
 }
 
-impl SendControlStream {
-    pub(super) fn new(conn: ConnectionRef) -> Self {
+impl SendUni {
+    pub(super) fn new(ty: StreamType, quic: quinn::Connection) -> Self {
         Self {
-            state: SendControlStreamState::Opening(conn.quic.open_uni()),
-            send: None,
-            conn,
+            ty,
+            state: SendUniState::New(quic),
+            data: VecDeque::with_capacity(2),
         }
     }
 }
 
-enum SendControlStreamState {
+enum SendUniState {
+    New(quinn::Connection),
```

Comment:
What is this `New` state for? It seems a bit iffy to me to have this contain a `quic::Connection`.

Reply:
This state reflects the moment when one SendStream has not been used. Depending on Peer's settings, it could never get to be used. I was hesitating between:

1. storing an OpenUni and not polling it until there is some data: The problem here is that we rely on the fact that OpenUni does not open a stream until it's polled.

2. having `ConnectionInner::send_uni` be `[ Option<SendUni>, 3 ]`, which would get some parts of the `SendUni` state machine live in `ConnectionInner`, and require a `quinn::Connection` back again in `ConnectionInner`.

I think `1.` would be the best solution ?

Reply:
Have going for `1.` in e16bfa9  H3: simplify SendUni state machine